include/envoy/request_info/dynamic_metadata.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <atomic>
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace RequestInfo {
+
+class DynamicMetadata {
+public:
+  virtual ~DynamicMetadata(){};
+
+  /**
+   * @param data_name the name of the data being set.
+   * @param data an owning pointer to the data to be stored.
+   * Note that it is an error to call setData() twice with the same data_name; this is to
+   * enforce a single authoritative source for each piece of data stored in DynamicMetadata.
+   */
+  template <typename T> void setData(absl::string_view data_name, std::unique_ptr<T>&& data) {
+    setDataGeneric(data_name, Traits<T>::getTypeId(), static_cast<void*>(data.release()),
+                   &Traits<T>::destructor);
+  }
+
+  /**
+   * @param data_name the name of the data being set.
+   * @return a reference to the stored data.
+   * Note that it is an error to access data that has not previously been set.
+   */
+  template <typename T> const T& getData(absl::string_view data_name) const {
+    return *static_cast<T*>(getDataGeneric(data_name, Traits<T>::getTypeId()));
```

Comment:
Did you consider adding an abstract empty class `DynamicMetadataObject` that all eligible metadata object inherit from? That way, you could us a safer `dynamic_cast` here to catch type violations via RTTI. This is what is done for TLS, see https://github.com/envoyproxy/envoy/blob/master/include/envoy/thread_local/thread_local.h#L16 and its uses (https://github.com/envoyproxy/envoy/blob/master/include/envoy/thread_local/thread_local.h#L41).